lib/_stream_transform.js

```diff
@@ -207,9 +207,7 @@ function done(stream, er, data) {
   if (data != null) // Single equals check for both `null` and `undefined`
     stream.push(data);
 
-  // TODO(BridgeAR): Write a test for these two error cases
-  // if there's nothing in the write buffer, then that means
-  // that nothing more will ever be provided
+  // These two error cases are sanity checks that can likely not be tested.
```

Comment:
Does this mean that the errors should be impossible to trigger? Or simply difficult to test? If they are impossible to trigger, let's change them to `assert()` instead?

Reply:
We should not depend on assert() in readable-stream (bundle size, etc), and I would need to spend time to actually remove it with regexps. So, can we avoid adding it in the first place?

Reply:
> We should not depend on assert() in readable-stream (bundle size, etc), and I would need to spend time to actually remove it with regexps. So, can we avoid adding it in the first place?

Yeah, I wasn't meaning `lib/assert.js` but rather the tiny (15 LoC) `lib/internal/assert.js` (that conditionally loads `lib/assert.js` only if it is actually throwing).  But I guess that doesn't help `readable-stream`?

Reply:
not at all :/.

Reply:
OK, in that case, how about just clearing up the comment to indicate whether this is something that is believed to logically be impossible (that is: we believe this error should never happen) or if it's just something that will indeed happen from time to time (due to bugs in people's code or problems on a host or whatever) and is merely difficult/impossible to reliably test.

Reply:
@Trott 
> If impossible, let's change them to assert() instead?

I originally planned exactly that: https://github.com/nodejs/node/pull/27073

> that conditionally loads lib/assert.js only if it is actually throwing

That would be changed if https://github.com/nodejs/node/pull/26635 lands. In that case it would be a regular internal error.

> merely difficult/impossible to reliably test

I believe this error should never happen. I am open for alternative wordings than the current one.

Reply:
`ERR_ASSERTION` can be imported from the same file as `ERR_TRANSFORM_WITH_LENGTH_0` and  `ERR_TRANSFORM_ALREADY_TRANSFORMING`. Does it make sense to change the lines below to throw `ERR_ASSERTION`? Or does that create similar problems for @mcollina as using `assert()`?

That would make it clear to the end user that they've found a bug in Node.js and should report it rather than that there's a bug in their own code. Then the comment could say something like: "Not using the assert module here because it creates problems for the readable-stream module. See https://github.com/nodejs/node/pull/27086/files/739dc8ef444102286be05df60a129d6ac5b9c963#r272457880."

Reply:
I have ported the error system.

Reply:
@mcollina if you ported the error system, is it possible to directly use all existing error types? If so, this will now use an an internal assert implementation that is independent from anything that the actual `assert` module does.

Reply:
That should be feasible.
See https://github.com/nodejs/readable-stream/blob/master/errors.js for more details on what’s there.

Reply:
In that case my former PR (https://github.com/nodejs/node/pull/27073/files) would do exactly that (it would throw an `ERR_ASSERTION` error but not use anything from `assert`).

Reply:
(`internal/assert` has nothing to do with `assert` anymore. They are 100% independent)

Reply:
I highly prefer not to, because it will make it harder to track future changes to that file as well. Adding another require could lead to 1-3 days of added work, and I prefer to be conservative for what is currently ok.

Reply:
Ok, I struggle what to do here in this case. @Trott are you fine with this comment due to the issues with `readable-streams`?

Reply:
> Ok, I struggle what to do here in this case. @Trott are you fine with this comment due to the issues with `readable-streams`?

Yeah, consider my comments to be nits. Would love to have consensus on something that makes everyone happy and addresses my comments/questions, but ultimately, I'm fine with whatever you and Matteo can agree on.

Reply:
Another alternative would be to remove these checks completely.
@mcollina are you fine with that?